File: python/Figure11_4.py
# Author: Kenji Kashima
# Date  : 2025/09/01
import numpy as np
import matplotlib.pyplot as plt
import config

np.random.seed(23)

def figure11_4(N_k= 100,label="a"):
    """
    Runs a SGD simulation for mean estimation.

    Parameters:
        N_k : simulation time length.
    """
    figsize = config.global_config(type= 1)
    C = [0.6, 0.6, 1.0, 1.0]        # constants for step size
    alpha = [1.0, 0.3, 1.0, 1.5]    # decay rate for step size
    N_setting = len(C)              # Number of settings

    p_ini = 1                       # initial value for SGD
    p_list = np.zeros((N_setting,N_k+1))
    y_list = np.zeros((N_setting,N_k))
    plt.figure(figsize=figsize)  
    p_list[:,0] = np.ones_like(p_list[:,0]) * p_ini 
    for setting in range( N_setting ):
        for k in range(N_k):        
            p = p_list[setting,k]
            z = np.random.randn()   # sampling of z_k 
            y_list[setting,k] = p - z   # Fig 11.4   
            # y_list[setting,i] = ( p > z ) - 0.5    # 演習 11.5 
            p_list[setting,k+1] = p - C[setting]/((k+1)**alpha[setting]) * y_list[setting,k]
        plt.plot(p_list[setting,:],label=r"$C={}, \alpha={}$".format(C[setting],alpha[setting]))

    plt.scatter(0,p_ini,marker='o',s=80,clip_on=False,color='black',label= "Initial Value",zorder=20)
    plt.ylabel(r"$p_k$")
    plt.xlabel(r"$k$")
    plt.xlim([0,N_k])
    plt.ylim([-2,2])
    plt.legend()
    plt.grid()
    plt.tight_layout()
    plt.savefig("./Figure11_4{}.pdf".format(label))
    plt.show()

if __name__ == '__main__':
    figure11_4(N_k=100,label="a")   
    figure11_4(N_k=5000,label="b") 

File: matlab/Figure11_4.m
% Author: Kenji Kashima
% Date  : 2025/06/05

clear; close all;
rng(23);  % Random seed

% Step‐size parameters
C_vals     = [0.6, 0.6, 1.0, 1.0];
alpha_vals = [1.0, 0.3, 1.0, 1.5];
numSettings = numel(C_vals);

% Two simulation lengths: (a) k_bar=100, (b) k_bar=5000
k_bar_list = [100, 5000];

for idxFig = 1:2
    k_bar = k_bar_list(idxFig);
    p_history = zeros(numSettings, k_bar+1);
    p_history(:,1) = 1;  % initial value p(0)=1 for all settings

    for s = 1:numSettings
        C = C_vals(s);
        alpha = alpha_vals(s);
        for k = 1:k_bar
            p_k      = p_history(s,k);
            z_k      = randn;                    % z_k ~ N(0,1)
            p_history(s,k+1) = p_k - C/((k)^alpha) * (p_k - z_k);
        end
    end

    % Plot all trajectories in one figure
    figure('Name', sprintf('Figure 11.4(%s)', char('a'+idxFig-1))); hold on; grid on;
    k_axis = 0:k_bar;
    colors = lines(numSettings);

    for s = 1:numSettings
        plot(k_axis, p_history(s,:), 'LineWidth', 1.5, 'Color', colors(s,:));
    end

    % Mark initial value at k=0
    scatter(0, 1, 80, 'k', 'filled', 'DisplayName', 'Initial Value');

    xlim([0, k_bar]);
    ylim([-2, 2]);
    xlabel('k', 'Interpreter','latex', 'FontSize',14);
    ylabel('$p_k$', 'Interpreter','latex', 'FontSize',14);

    % Build legend entries for the four step‐size settings
    legendEntries = cell(1, numSettings+1);
    for s = 1:numSettings
        legendEntries{s} = sprintf('$C=%.1f,\\ \\alpha=%.1f$', C_vals(s), alpha_vals(s));
    end
    legendEntries{end} = 'Initial Value';

    legend(legendEntries, 'Interpreter','latex', 'FontSize',10, 'Location','best');
    hold off;
end
